src/components/tl/ucp/allgather/allgather_neighbor.c

```diff
@@ -0,0 +1,176 @@
+#include "config.h"
+#include "tl_ucp.h"
+#include "allgather.h"
+#include "core/ucc_progress_queue.h"
+#include "tl_ucp_sendrecv.h"
+#include "utils/ucc_math.h"
+#include "utils/ucc_coll_utils.h"
+#include "components/mc/ucc_mc.h"
+
+static ucc_rank_t get_recv_from_rank(ucc_rank_t rank, ucc_rank_t size, int i)
```

Comment:
add pointer to neighbors array as a param of function so not to recalculate

Reply:
just simplified it

Reply:
what about defining an auxiliary function like that:

```
static inline ucc_rank_t offset_at_step(int i, ucc_rank_t rank) {
   return 2 * ((i % 2) ^^ (rank % 2) ? 1 : -1);
}
```

Reply:
nice, but this is task for compiler to optimize this expression in such way, I would prefer to leave as is to be more human readable. (don't think it is hotspot anyway)